python/orca/src/bigdl/orca/data/shard.py

```diff
@@ -542,18 +549,64 @@ def f(iter):
                 np_records = pdf.to_records(index=False)
                 return [r.tolist() for r in np_records]
 
-        def getSchema(iter):
-            for pdf in iter:
-                return [pdf.columns.values]
-
         rdd = self.rdd.mapPartitions(f)
-        column = self.rdd.mapPartitions(getSchema).first()
+        column = self.get_schema()['columns']
         df = rdd.toDF(list(column))
-        df.cache()
-        df.count()
-        self.uncache()
         return df
 
+    def to_spark_df(self):
+        if self._get_class_name() != 'pandas.core.frame.DataFrame':
+            invalidInputError(False,
+                              "Currently only support to_spark_df on XShards of Pandas DataFrame")
+
+        try:
+            def getSchemaStructType(iter):
+                for pdf in iter:
+                    schema = [str(x) if not isinstance(x, str) else x for x in pdf.columns]
+                    if isinstance(schema, (list, tuple)):
+                        arrow_schema = pa.Schema.from_pandas(pdf, preserve_index=False)
+                        struct = StructType()
+                        for name, field in zip(schema, arrow_schema):
+                            struct.add(
+                                name, from_arrow_type(field.type), nullable=field.nullable
+                            )
+                        schema = struct
+                        return [schema]
+
+            schema = self.rdd.mapPartitions(getSchemaStructType).first()
```

Comment:
van you use `self.get_schema()` to get the pandas schema instead of `mapPartitions`?

Reply:
`mapPartitions` will return pyspark schema(`structtype`) which is needed to create df

Reply:
> `mapPartitions` will return pyspark schema(`structtype`) which is needed to create spark df

You may create pyspark schema `structtype` from the pandas schema returned by `self.get_schema()`, yes?

Reply:
'self.get_schema()' takes more time than create `structtype` by `mapPartitions`, 0.37s VS 0.2 s, if we create `structtype` from 'self.get_schema()', it will take longer.

Reply:
> 'self.get_schema()' takes more time than create `structtype` by `mapPartitions`, 0.37s VS 0.2 s, if we create `structtype` from 'self.get_schema()', it will take longer.

How's 'self.get_schema()' implemented? You can fix that method.